src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala

```diff
@@ -293,18 +279,91 @@ object Utils extends Logging {
 
   var eid = 0L
   var attested: Boolean = false
-  var attesting_getepid: Boolean = false
-  var attesting_getmsg1: Boolean = false
-  var attesting_getmsg3: Boolean = false
-  var attesting_final_ra: Boolean = false
+  // Initialize accumulator variables for tracking the state of attestation
+  var acc_registered: Boolean = false
+  val numEnclaves: LongAccumulator = new LongAccumulator
+  val numAttested: LongAccumulator = new LongAccumulator
+  var loop : Boolean = true
 
   def initSQLContext(sqlContext: SQLContext): Unit = {
     sqlContext.experimental.extraOptimizations =
       (Seq(EncryptLocalRelation, ConvertToOpaqueOperators) ++
         sqlContext.experimental.extraOptimizations)
     sqlContext.experimental.extraStrategies = (Seq(OpaqueOperators) ++
       sqlContext.experimental.extraStrategies)
-    RA.initRA(sqlContext.sparkContext)
+
+    val sc = sqlContext.sparkContext
+    // This is needed to prevent an error from re-registering accumulator variables if
+    // `initSQLContext` is called multiple times on the driver
+    if (!acc_registered) {
+      sc.register(numEnclaves)
+      sc.register(numAttested)
+      acc_registered = true
+    }
+
+    RA.waitForExecutors(sc)
+    RA.attestEnclaves(sc)
+    RA.startThread(sc)
+  }
+
+  def initEnclave(): (SGXEnclave, Long) = {
+    // This following exception relies on the Spark fault tolerance and its retry mechanism
+    // in case of a task failure. Therefore, Spark MUST be configured to retry in case of a task failure.
+    this.synchronized {
+      if (!attested) {
+        Thread.sleep(200)
+        throw new OpaqueException("Attestation not yet complete")
+      } else {
+        val enclave = new SGXEnclave()
+        (enclave, eid)
+      }
+    }
+  }
+
+  def startEnclave(numEnclavesAcc: LongAccumulator) : Long = {
+    this.synchronized {
+      if (eid == 0L) {
+        val enclave = new SGXEnclave()
+        val path = findLibraryAsResource("enclave_trusted_signed")
+        eid = enclave.StartEnclave(path)
+        numEnclavesAcc.add(1)
+        logInfo("Starting an enclave")
+      }
+    }
+    eid
+  }
+
+  def generateReport(): (Long, Option[Array[Byte]]) = {
+    this.synchronized {
+      // Only generate report if the enclave has already been started, but not yet attested
+      if (eid != 0L && !attested) {
+        val enclave = new SGXEnclave()
+        val msg1 = enclave.GenerateReport(eid)
+        (eid, Option(msg1))
+      } else {
+        (eid, None)
+      }
+    }
+  }
+
+  def finishAttestation(
+      numAttested: LongAccumulator,
+      msg2s: Map[Long, Array[Byte]]
+  ): (Long, Boolean) = {
+    this.synchronized {
+      val enclave = new SGXEnclave()
+      if (msg2s.contains(eid) && !attested) {
+        val msg2 = msg2s(eid)
+        enclave.FinishAttestation(eid, msg2)
+        numAttested.add(1)
+        attested = true
+      }
+      (eid, attested)
+    }
+  }
+
+  def cleanup() {
```

Comment:
I would recommend include `spark.stop()` here so you don't have to always couple those two functions together.